File: basics/bst-llrb.go
package main

import (
	"fmt"
)

type Interface interface {
	Less(Interface) bool
}

const (
	black = false
	red = true
)

type color bool

func (c color) String() string {
	if c {
		return "Red"
	}

	return "Black"
}

type tree struct {
	key	Interface
	left	*tree
	right	*tree
	color
}

func (t *tree) String() string {
	return fmt.Sprintf("k:%v c:%v", t.key, t.color)
}

func (t *tree) RotateLeft() *tree {
	s := t.right
	t.right = s.left
	s.left = t

	s.color = t.color
	t.color = red

	return s
}

func (t *tree) RotateRight() *tree {
	s := t.left
	t.left = s.right
	s.right = t

	s.color = t.color
	t.color = red

	return s
}

func (t *tree) FlipColor() {
	t.color = !t.color
}

func (t *tree) FlipColors() {
	t.FlipColor()
	t.left.FlipColor()
	t.right.FlipColor()
}

func (t *tree) IsRed() bool {
	return t != nil && bool(t.color)
}

func (t *tree) MoveRedLeft() *tree {
	t.FlipColors()

	if t.right.left.IsRed() {
		t.right = t.right.RotateRight()
		t = t.RotateLeft()
		t.FlipColors()
	}

	return t
}

func (t *tree) MoveRedRight() *tree {
	t.FlipColors()

	if t.left.left.IsRed() {
		t = t.RotateLeft()
		t.FlipColors()
	}

	return t
}

func (t *tree) FixUp() *tree {
	if t.right.IsRed() {
		t = t.RotateLeft()
	}
	if t.left.IsRed() && t.left.left.IsRed() {
		t = t.RotateRight()
	}
	if t.left.IsRed() && t.right.IsRed() {
		t.FlipColors()
	}
	return t
}

func (t *tree) Min() Interface {
	for t.left != nil {
		t = t.left
	}

	return t.key
}

func (t *tree) DeleteMin() *tree {
	if t.left == nil {
		return nil
	}

	if !t.left.IsRed() && !t.left.left.IsRed() {
		t = t.MoveRedLeft()
	}

	t.left = t.left.DeleteMin()

	return t.FixUp()
}

func (t *tree) Insert(k Interface) *tree {
	if t == nil {
		return &tree{key: k, color: red}
	}

	if k.Less(t.key) {
		t.left = t.left.Insert(k)
	} else {
		t.right = t.right.Insert(k)
	}

	return t.FixUp()
}

func (t *tree) Delete(k Interface) *tree {
	if k.Less(t.key) {
		if !t.left.IsRed() && !t.left.left.IsRed() {
			t = t.MoveRedLeft()
		}

		t.left = t.left.Delete(k)
	} else {
		if t.left.IsRed() {
			t = t.RotateRight()
		}

		if k == t.key && t.right == nil {
			return nil
		}
		if !t.right.IsRed() && !t.right.left.IsRed() {
			t = t.MoveRedRight()
		}

		if k == t.key {
			t.key = t.right.Min()
			t.right = t.right.DeleteMin()
		} else {
			t.right = t.right.Delete(k)
		}
	}

	return t.FixUp()
}

type Tree struct { root *tree }

func (t *Tree) Insert(k Interface) {
	t.root = t.root.Insert(k)
	t.root.color = black
}

func (t *Tree) Find(k Interface) bool {
	s := t.root
	for s != nil && s.key != k {
		if k.Less(s.key) {
			s = s.left
		} else {
			s = s.right
		}
	}

	return s != nil
}

func (t *Tree) Delete(k Interface) {
	t.root = t.root.Delete(k)
	if t.root != nil {
		t.root.color = black
	}
}

func (t *tree) Print(d int) {
	if t != nil {
		t.left.Print(d + 1)
		fmt.Printf("%v:%d:%v ", t.key, d, t.color)
		t.right.Print(d + 1)
	}
}

func (t *Tree) Print() {
	t.root.Print(1)
}

type Int int

func (i Int) Less(j Interface) bool {
	return i < j.(Int)
}

func main() {
	t := &Tree{}

	for {
		var c string
		fmt.Scanf("%s", &c)

		var v Int
		if c != "p" {
			fmt.Scanf("%d", &v)
		}

		switch c {
		case "i":
			t.Insert(v)
		case "d":
			t.Delete(v)
		case "f":
			fmt.Println("find: ", t.Find(v))
		case "p":
			t.Print()
			fmt.Println()
		}
	}
}

File: basics/bst-llrb.c
#include <stdio.h>

int f, root, k[1000], c[1000], l[1000], r[1000];

int rotate_left(int i) {
	int j = r[i]; r[i] = l[j]; l[j] = i;
	c[j] = c[i]; c[i] = 1;
	return j;
}

int rotate_right(int i) {
	int j = l[i]; l[i] = r[j]; r[j] = i;
	c[j] = c[i]; c[i] = 1;
	return j;
}

void flip_colors(int i) {
	c[i] = !c[i]; c[l[i]] = !c[l[i]]; c[r[i]] = !c[r[i]];
}

int is_red(int i) {
	return i != -1 && c[i] == 1;
}

void insert(int x) {
	int insert_rec(int i) {
		if (i == -1) {
			i = f; f++;
			k[i] = x; l[i] = r[i] = -1; c[i] = 1; 
			return i;
		}

		if (x < k[i])
			l[i] = insert_rec(l[i]);
		else
			r[i] = insert_rec(r[i]);

		if (is_red(r[i]) && !is_red(l[i]))
			i = rotate_left(i);
		if (is_red(l[i]) && is_red(l[l[i]]))
			i = rotate_right(i);
		if (is_red(l[i]) && is_red(r[i]))
			flip_colors(i);

		return i;
	}

	root = insert_rec(root);
	c[root] = 0;
}

int find(int x) {
	int i = root;

	while (i != -1 && k[i] != x)
		if (x < k[i])
			i = l[i];
		else
			i = r[i];

	return i != -1;
}

int move_red_left(int i) {
	flip_colors(i);

	if (is_red(l[r[i]])) {
		r[i] = rotate_right(r[i]);
		i = rotate_left(i);
		flip_colors(i);
	}

	return i;
}

int move_red_right(int i) {
	flip_colors(i);

	if (is_red(l[l[i]])) {
		i = rotate_left(i);
		flip_colors(i);
	}

	return i;
}

int fix_up(int i) {
	if (is_red(r[i]))
		i = rotate_left(i);
	if (is_red(l[i]) && is_red(l[l[i]]))
		i = rotate_right(i);
	if (is_red(l[i]) && is_red(r[i]))
		flip_colors(i);

	return i;
}

int min_rec(int i) {
	while (l[i] != -1)
		i = l[i];
	return k[i];
}

int delete_min_rec(int i) {
	if (l[i] == -1)
		return -1;

	if (!is_red(l[i]) && !is_red(l[l[i]]))
		i = move_red_left(i);

	l[i] = delete_min_rec(l[i]);

	return fix_up(i);
}

void delete(int x) {
	int delete_rec(int i) {
		if (x < k[i]) {
			if (!is_red(l[i]) && !is_red(l[l[i]]))
				i = move_red_left(i);
			l[i] = delete_rec(l[i]);
		} else {
			if (is_red(l[i]))
				i = rotate_right(i);
			
			if (x == k[i] && r[i] == -1) {
				return -1;
			}

			if (!is_red(r[i]) && !is_red(l[r[i]]))
				i = move_red_right(i);

			if (x == k[i]) {
				k[i] = min_rec(r[i]);
				r[i] = delete_min_rec(r[i]);
			} else
				r[i] = delete_rec(r[i]);
		}

		return fix_up(i);
	}

	root = delete_rec(root);
	c[root] = 0;
}

void print() {
	void print_rec(int i) {
		if (i != -1) {
			print_rec(l[i]);
			printf("%d:%c ", k[i], c[i] ? 'r' : 'b');
			print_rec(r[i]);
		}
	}

	print_rec(root);
	putchar('\n');
}

int main() {
	char c;
	int x;

	f = 0; root = -1;

	while (1) {
		scanf("%c", &c);
		if (c != 'p')
			scanf("%d", &x);

		switch (c) {
			case 'i':
				insert(x);
				break;
			case 'd':
				delete(x);
				break;
			case 'f':
				printf("find: %d\n", find(x));
				break;
			case 'p':
				print();
		}
	}
}
